horovod/tensorflow/__init__.py

```diff
@@ -304,13 +273,69 @@ def DistributedGradientTape(gradtape, device_dense='', device_sparse='',
+        import horovod.tensorflow.keras as hvd_k
+        return hvd_k.DistributedOptimizer(optimizer, name, device_dense, device_sparse,
+                                          compression, sparse_as_dense)
     else:
-        return cls(gradtape._tape, device_dense, device_sparse,
-                   compression, sparse_as_dense,
-                   gradtape._persistent)
+        raise ValueError('Provided optimizer doesn\'t inherit from either legacy '
+                         'TensorFlow or Keras optimizer: %s' % optimizer)
+
+
+if hasattr(tf, 'GradientTape'):
+    class _DistributedGradientTape(tf.GradientTape):
+        def __init__(self, tape, device_dense, device_sparse, compression, sparse_as_dense,
+                     persistent=False, watch_accessed_variables=True):
+            if hasattr(tape, '_watch_accessed_variables'):
+                super(self.__class__, self).__init__(persistent, watch_accessed_variables)
+            else:
+                super(self.__class__, self).__init__(persistent)
+
+            self._tape = tape
+            self._allreduce_grads = _make_allreduce_grads_fn(
+                'DistributedGradientTape', device_dense, device_sparse, compression,
+                sparse_as_dense)
+
+        def gradient(self, target, sources, output_gradients=None):
+            gradients = super(self.__class__, self).gradient(target, sources, output_gradients)
+            if size() > 1:
+                return self._allreduce_grads(gradients)
+            else:
+                return gradients
+
+
+    def DistributedGradientTape(gradtape, device_dense='', device_sparse='',
+                                compression=Compression.none, sparse_as_dense=False):
+        """An tape that wraps another tf.GradientTape, using an allreduce to
+        average gradient values before applying gradients to model weights.
+
+        Args:
+          gradtape:
+            GradientTape to use for computing gradients and applying updates.
+          device_dense:
+            Device to be used for dense tensors. Uses GPU by default
+            if Horovod was build with HOROVOD_GPU_ALLREDUCE.
+          device_sparse:
+            Device to be used for sparse tensors. Uses GPU by default
+            if Horovod was build with HOROVOD_GPU_ALLGATHER.
+          compression:
+            Compression algorithm used during allreduce to reduce the amount
+            of data sent during the each parameter update step.  Defaults to
+            not using compression.
+          sparse_as_dense:
+            Treat all sparse gradients as dense tensors.  This can help improve
+            performance and memory utilization if the original sparse gradient
+            has high density.  Defaults to false.
+        """
+        cls = type(gradtape.__class__.__name__, (gradtape.__class__,),
+                   dict(_DistributedGradientTape.__dict__))
+        if hasattr(gradtape, '_watch_accessed_variables'):
```

Comment:
Please comment on why this check is necessary.

Reply:
I'm actually not sure, it was introduced in #670.